core/worker.go

```diff
@@ -598,18 +598,81 @@ func (w *worker) GeneratePendingHeader(block *types.WorkObject, fill bool, txs t
 
 		// Add an etx for each workshare for it to be rewarded
 		for _, uncle := range uncles {
-			reward := misc.CalculateReward(uncle)
+			reward := misc.CalculateReward(block, uncle)
 			uncleCoinbase := uncle.PrimaryCoinbase()
 			work.etxs = append(work.etxs, types.NewTx(&types.ExternalTx{To: &uncleCoinbase, Gas: params.TxGas, Value: reward, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(work.etxs)), Sender: uncleCoinbase, Data: []byte{uncle.Lock()}}))
 		}
 
 	}
 
+	if nodeCtx == common.ZONE_CTX {
+		var exchangeRate *big.Int
+		if w.hc.IsGenesisHash(block.Hash()) {
+			exchangeRate = params.ExchangeRate
+		} else {
+			if block.NumberU64(common.ZONE_CTX) > params.ControllerKickInBlock {
+				// convert map to a slice
+				updatedTokenChoiceSet, err := CalculateTokenChoicesSet(w.hc, block, work.etxs)
+				if err != nil {
+					return nil, err
+				}
+				exchangeRate, _, _, err = CalculateExchangeRate(w.hc, block, updatedTokenChoiceSet)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				exchangeRate = block.ExchangeRate()
+			}
+		}
+		work.wo.Header().SetExchangeRate(exchangeRate)
+
+		for _, etx := range work.etxs {
+			// If the etx is conversion
+			if types.IsConversionTx(etx) {
+				value := etx.Value()
+				// If sender is in Quai, convert the value into Qi
+				if etx.ETXSender().IsInQuaiLedgerScope() {
+					value = misc.QuaiToQi(work.wo, value)
+				}
+				// If sender is in Qi, convert the value into Quai
+				if etx.ETXSender().IsInQiLedgerScope() {
```

Comment:
Same issue, I don't think EtxSender is valid here

Reply:
changed it